### ISSUE TITLE
TimeoutException

### ISSUE THREAD

**Issue by colinbes** [alexandru/shade#15]
I am wondering how to ensure graceful handling of timeout exception when get or set of memcache when memcache is not actually running.

I must be doing something wrong in trying to capture the exception and would appreciate any advice.

I am assuming I need to wrap the memcache.get with a try/catch block.

**Comment by colinbes** [comment]
I ended up modifying my code - I thought I had to catch exception on the get/set function as opposed to when future is processed. I ended up using Either method to gracefully handle failure in my calling code.
As example see FastStore, a convenience object for access Memcache

```
object FastStore {
  lazy val memcached: Memcached = Memcached(Configuration("127.0.0.1:11211"), ec)

  def get[T: shade.memcached.Codec](key: String):Future[Either[Throwable, Option[T]]] =  {
    memcached.get[T](key).map { x => 
      Right(x)
    } recover {
      case cause => Left(cause)} 
    }

  def set ..
}
```

**Comment by alexandru** [comment]
@colinbes yes, `recover` and `recoverWith` are the helpers you get on `Future` to handle exceptions, plus you can also handle those errors with a simple `.onComplete(callback)` or an `.onFailure(callback)`.

Btw, a `Future` is basically a delayed `Try` and a `Try[T]` is equivalent to an `Either[Throwable, T]`. There aren't many reasons for turning a `Future[T]` into a `Future[Either[Throwable, T]]`. But if you must, I would still use `Try` instead of either, meaning that I would turn `Future[T]` into a `Future[Try[T]]`. Here's a helper that I use ...

```scala
def liftTry[T](source: Future[T])(implicit ec: ExecutionContext): Future[Try[T]] = {
  val p = Promise[Try[T]]()
  source.onComplete { case result => p.success(result) }
  p.future
}
```

And in fact, because Shade currently depends on [monifu-core](https://github.com/monifu/monifu) (for the time being, though I'm thinking of removing it in the next version), you can just use this helper as it is defined in Monifu, like so:

```scala
import monifu.concurrent.extensions._

// now this works ...
val result: Future[Try[Option[T]]] =
  memcached.get[T](key).liftTry
```

**Comment by alexandru** [comment]
@colinbes also make sure to update to the latest version of Shade, I released 1.7.1 yesterday.

**Comment by colinbes** [comment]
Thanks @alexandru for the good input and the update!
